lib/std/smartptrs.nim

```diff
@@ -0,0 +1,212 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2021 Nim contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+
+## C++11 like smart pointers. They always use the shared allocator.
+
+import std/isolation
+
+
+type
+  UniquePtr*[T] = object
+    ## Non copyable pointer to a value of type `T` with exclusive ownership.
+    val: ptr T
+
+proc `=destroy`*[T](p: var UniquePtr[T]) =
+  if p.val != nil:
+    `=destroy`(p.val[])
+    when compileOption("threads"):
+      deallocShared(p.val)
+    else:
+      dealloc(p.val)
+
+proc `=`*[T](dest: var UniquePtr[T], src: UniquePtr[T]) {.error.}
+  ## The copy operation is disallowed for `UniquePtr`, it
+  ## can only be moved.
+
+proc newUniquePtr*[T](val: sink Isolated[T]): UniquePtr[T] {.nodestroy.} =
```

Comment:
```suggestion
proc initUniquePtr*[T](val: sink Isolated[T]): UniquePtr[T] {.nodestroy.} =
```

new is for ref types.

(furthermore, `isNil` would be confusing for humans otherwise, as they'd assume it'd check whether the ref is nil, whereas it's instead checking for get)

Reply:
However, these are pointer types (which are similar to ref types), so it makes sense to use `new` imo.

> furthermore, `isNil` would be confusing for humans otherwise, as they'd assume it'd check whether the ref is nil

It does check if the pointer is nil, so I'm not sure what you mean.

Reply:
see https://github.com/nim-lang/Nim/pull/17495